### PR TITLE
#450 rsync command doesn't work in windows with cygwin due to broken ControlMaster support

### PR DESCRIPTION
**Pull request by xunnanxu**

### CODE REVIEW

File: test/ssh_command_test.rb [hunk 1]
```diff
@@ -69,7 +75,7 @@ def test_try_connection_with_gateway_connects_using_ssh_gateway
     cmd = command("10.0.0.1", "--ssh-gateway=user@gateway")
     ssh_mock = mock 'ssh_mock'
     Net::SSH::Gateway.expects(:new).with('gateway', 'user').returns(ssh_mock)
-    ssh_mock.expects(:ssh).with(cmd.host, cmd.user, cmd.connection_options.except(:gateway))
+    ssh_mock.expects(:ssh).with(cmd.host, cmd.user, cmd.connection_options.select{|o| o != :gateway})
```
Comment (matschaffer) [inline test/ssh_command_test.rb]: Why the change here? Could I get some info?

Comment (xunnanxu) [inline test/ssh_command_test.rb]: I think hash.except is something in rails? Does ruby support that itself?

I got this when I ran the test:
>SshCommandTest#test_try_connection_with_gateway_connects_using_ssh_gateway:
NoMethodError: undefined method `except' for #<Hash:0x00000601f0d7d0>

Comment (matschaffer) [inline test/ssh_command_test.rb]: Oh odd. This test was passing. Wonder if we were getting the method from a
transient gem.

Thanks then!

On Saturday, September 12, 2015, Shawn Xu <notifications@github.com> wrote:

> In test/ssh_command_test.rb
> <https://github.com/matschaffer/knife-solo/pull/453#discussion_r39331141>:
>
> > @@ -69,7 +75,7 @@ def test_try_connection_with_gateway_connects_using_ssh_gateway
> >      cmd = command("10.0.0.1", "--ssh-gateway=user@gateway")
> >      ssh_mock = mock 'ssh_mock'
> >      Net::SSH::Gateway.expects(:new).with('gateway', 'user').returns(ssh_mock)
> > -    ssh_mock.expects(:ssh).with(cmd.host, cmd.user, cmd.connection_options.except(:gateway))
> > +    ssh_mock.expects(:ssh).with(cmd.host, cmd.user, cmd.connection_options.select{|o| o != :gateway})
>
> I think hash.except is something in rails? Does ruby support that itself?
>
> I got this when I ran the test:
>
> SshCommandTest#test_try_connection_with_gateway_connects_using_ssh_gateway:
> NoMethodError: undefined method `except' for #Hash:0x00000601f0d7d0
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matschaffer/knife-solo/pull/453/files#r39331141>.
>


-- 

-Mat

matschaffer.com

Comment (xunnanxu) [inline test/ssh_command_test.rb]: Yeah. You can definitely get it from ActiveSupport but I don't see anything requiring it in this repo so I'm not sure how it used to work. It certainly works when you run all the tests but fails if you just run this one.

Comment (matschaffer) [inline test/ssh_command_test.rb]: Ah, it's coming from buff-extensions via berkshelf. That would explain why it works running the full suite since there are berkshelf tests in there too. Thanks for finding this!

**Comment by matschaffer** [comment]
Thanks! I'll plan to get this out as a point release soon (hopefully this week).

I *might* pull out the control socket stuff for 0.6.0 since I think pre-staging assets might be simpler and open up parallelization options.

But definitely good to have this in the mean time.
